src/realm/alloc_slab.cpp

```diff
@@ -718,16 +718,9 @@ bool SlabAlloc::align_filesize_for_mmap(ref_type top_ref, Config& cfg)
     // check if online compaction allows us to shrink the file:
     if (top_ref) {
         // Get the expected file size by looking up logical file size stored in top array
-        constexpr size_t max_top_size = (Group::s_file_size_ndx + 1) * 8 + sizeof(Header);
-        size_t top_page_base = top_ref & ~(page_size() - 1);
-        size_t top_offset = top_ref - top_page_base;
-        size_t map_size = std::min(max_top_size + top_offset, size - top_page_base);
-        File::Map<char> map_top(m_file, top_page_base, File::access_ReadOnly, map_size, 0, m_write_observer);
-        realm::util::encryption_read_barrier(map_top, top_offset, max_top_size);
```

Comment:
Here we were requesting the `max_top_size` which may be more than the encrypted mappings hold (or the file for that matter) if the top_ref happened to be very close to the end of the file. If the "top ref + size" crossed a page_size, then the File::Map won't have the mappings set up, and we'll hit an assertion. If the "top ref + size" crossed the end of the file, then we'd get a DecryptionFailed exception. 

All of this is rather tricky to calculate accurately because we'd first have to map/decrypt the top_ref header, and then based on the array width, calculate the number of bytes to the `Group::s_file_size_ndx` position and then map/decrypt that before we can read the data. Fortunately all of this is already handled by standard code to read from an array, so we can just use that existing machinery. This also means that we'd use our existing mappings rather than creating a new one here just for this.